pkg/controller/scheduler.go

```diff
@@ -446,13 +447,34 @@ func (c *Controller) advanceCanary(name string, namespace string, skipLivenessCh
 
 	// canary incremental traffic weight
 	if canaryWeight < maxWeight {
-		primaryWeight -= cd.Spec.CanaryAnalysis.StepWeight
-		if primaryWeight < 0 {
-			primaryWeight = 0
-		}
-		canaryWeight += cd.Spec.CanaryAnalysis.StepWeight
-		if primaryWeight > 100 {
-			primaryWeight = 100
+		// If in "mirror" mode, do one step of mirroring before shifting traffic to canary.
+		// When mirroring, all requests go to primary and canary, but only responses from
+		// primary go back to the user.
+		c.logger.With("canary", fmt.Sprintf("%s.%s", name, namespace)).
+			Warnf("Deciding on canary, input: %d/%d/%t, Mirror: %t",
+				primaryWeight, canaryWeight, mirrored, cd.Spec.CanaryAnalysis.Mirror)
+		if cd.Spec.CanaryAnalysis.Mirror && canaryWeight == 0 {
+			if mirrored == false {
+				mirrored = true
+				primaryWeight = 100
+				canaryWeight = 0
+			} else {
+				mirrored = false
+				primaryWeight = 100 - cd.Spec.CanaryAnalysis.StepWeight
+				canaryWeight = cd.Spec.CanaryAnalysis.StepWeight
+			}
+			c.logger.With("canary", fmt.Sprintf("%s.%s", name, namespace)).
+				Warnf("Running mirror step %d/%d/%t", primaryWeight, canaryWeight, mirrored)
```

Comment:
I think this should be `Infof` instead of `Warnf `.

Reply:
Yes sorry I think there are a few log messages that may be extraneous or at too high of a level.  Let me do a pass for all logs I added.